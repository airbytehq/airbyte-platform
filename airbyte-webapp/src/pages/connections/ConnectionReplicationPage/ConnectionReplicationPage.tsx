import React, { useCallback, useEffect } from "react";
import { useFormState } from "react-hook-form";
import { FormattedMessage, useIntl } from "react-intl";
import { useLocation } from "react-router-dom";
import { useUnmount } from "react-use";

import { ConnectionConfigurationCard } from "components/connection/ConnectionForm/ConnectionConfigurationCard";
import {
  FormConnectionFormValues,
  useConnectionValidationSchema,
} from "components/connection/ConnectionForm/formConfig";
import { useRefreshSourceSchemaWithConfirmationOnDirty } from "components/connection/ConnectionForm/refreshSourceSchemaWithConfirmationOnDirty";
import { SchemaChangeBackdrop } from "components/connection/ConnectionForm/SchemaChangeBackdrop";
import { SyncCatalogCard } from "components/connection/ConnectionForm/SyncCatalogCard";
import { UpdateConnectionFormControls } from "components/connection/ConnectionForm/UpdateConnectionFormControls";
import { SchemaError } from "components/connection/CreateConnectionForm/SchemaError";
import { compareObjectsByFields } from "components/connection/syncCatalog/utils";
import { Form } from "components/forms";
import LoadingSchema from "components/LoadingSchema";
import { FlexContainer } from "components/ui/Flex";
import { Message } from "components/ui/Message/Message";

import { ConnectionValues, useGetStateTypeQuery } from "core/api";
import {
  AirbyteStreamAndConfiguration,
  AirbyteStreamConfiguration,
  WebBackendConnectionRead,
  WebBackendConnectionUpdate,
} from "core/api/types/AirbyteClient";
import {
  getFrequencyFromScheduleData,
  Action,
  Namespace,
  PageTrackingCodes,
  useAnalyticsService,
  useTrackPage,
} from "core/services/analytics";
import { equal } from "core/utils/objects";
import { useConfirmCatalogDiff } from "hooks/connection/useConfirmCatalogDiff";
import { useSchemaChanges } from "hooks/connection/useSchemaChanges";
import { useConnectionEditService } from "hooks/services/ConnectionEdit/ConnectionEditService";
import { useConnectionFormService } from "hooks/services/ConnectionForm/ConnectionFormService";
import { useModalService } from "hooks/services/Modal";

import styles from "./ConnectionReplicationPage.module.scss";
import { ResetWarningModal } from "./ResetWarningModal";

const toWebBackendConnectionUpdate = (connection: WebBackendConnectionRead): WebBackendConnectionUpdate => ({
  name: connection.name,
  connectionId: connection.connectionId,
  namespaceDefinition: connection.namespaceDefinition,
  namespaceFormat: connection.namespaceFormat,
  prefix: connection.prefix,
  syncCatalog: connection.syncCatalog,
  scheduleData: connection.scheduleData,
  scheduleType: connection.scheduleType,
  status: connection.status,
  resourceRequirements: connection.resourceRequirements,
  operations: connection.operations,
  sourceCatalogId: connection.catalogId,
});

const SchemaChangeMessage: React.FC = () => {
  const { isDirty } = useFormState<FormConnectionFormValues>();
  const refreshWithConfirm = useRefreshSourceSchemaWithConfirmationOnDirty(isDirty);

  const { refreshSchema } = useConnectionFormService();
  const { connection, schemaHasBeenRefreshed } = useConnectionEditService();
  const { hasNonBreakingSchemaChange, hasBreakingSchemaChange } = useSchemaChanges(connection.schemaChange);

  if (schemaHasBeenRefreshed) {
    return null;
  }

  if (hasNonBreakingSchemaChange) {
    return (
      <Message
        type="info"
        text={<FormattedMessage id="connection.schemaChange.nonBreaking" />}
        actionBtnText={<FormattedMessage id="connection.schemaChange.reviewAction" />}
        onAction={refreshSchema}
        data-testid="schemaChangesDetected"
      />
    );
  }

  if (hasBreakingSchemaChange) {
    return (
      <Message
        type="error"
        text={<FormattedMessage id="connection.schemaChange.breaking" />}
        actionBtnText={<FormattedMessage id="connection.schemaChange.reviewAction" />}
        onAction={refreshWithConfirm}
        data-testid="schemaChangesDetected"
      />
    );
  }
  return null;
};

export const ConnectionReplicationPage: React.FC = () => {
  const analyticsService = useAnalyticsService();
  useTrackPage(PageTrackingCodes.CONNECTIONS_ITEM_REPLICATION);

  const getStateType = useGetStateTypeQuery();

  const { formatMessage } = useIntl();
  const { openModal } = useModalService();

  const { connection, schemaRefreshing, updateConnection, discardRefreshedSchema } = useConnectionEditService();
  const { initialValues, schemaError, setSubmitError, refreshSchema, mode } = useConnectionFormService();
  const validationSchema = useConnectionValidationSchema();

  const saveConnection = useCallback(
    async (
      values: ConnectionValues,
      { skipReset, catalogHasChanged }: { skipReset: boolean; catalogHasChanged: boolean }
    ) => {
      const connectionAsUpdate = toWebBackendConnectionUpdate(connection);

      await updateConnection({
        ...connectionAsUpdate,
        ...values,
        connectionId: connection.connectionId,
        skipReset,
      });

      if (catalogHasChanged) {
        // TODO (https://github.com/airbytehq/airbyte/issues/17666): Move this into a useTrackChangedCatalog method (name pending) post Vlad's analytics hook work
        analyticsService.track(Namespace.CONNECTION, Action.EDIT_SCHEMA, {
          actionDescription: "Connection saved with catalog changes",
          connector_source: connection.source.sourceName,
          connector_source_definition_id: connection.source.sourceDefinitionId,
          connector_destination: connection.destination.destinationName,
          connector_destination_definition_id: connection.destination.destinationDefinitionId,
          frequency: getFrequencyFromScheduleData(connection.scheduleData),
        });
      }
    },
    [analyticsService, connection, updateConnection]
  );

  const onFormSubmit = useCallback(
    async (values: FormConnectionFormValues) => {
      // Check if the user refreshed the catalog and there was any change in a currently enabled stream
      const hasDiffInEnabledStream = connection.catalogDiff?.transforms.some(({ streamDescriptor }) => {
        // Find the stream for this transform in our form's syncCatalog
        const stream = values.syncCatalog.streams.find(
          ({ stream }) => streamDescriptor.name === stream?.name && streamDescriptor.namespace === stream.namespace
        );
        return stream?.config?.selected;
      });

      // Check if the user made any modifications to enabled streams compared to the ones in the latest connection
      // e.g. changed the sync mode of an enabled stream
      const hasUserChangesInEnabledStreams = !equal(
        values.syncCatalog.streams.filter((s) => s.config?.selected),
        connection.syncCatalog.streams.filter((s) => s.config?.selected)
      );

      // Only adding/removing a stream - with 0 other changes - doesn't require a reset
      // for each form value stream, find the corresponding connection value stream
      // and remove `config.selected` from both before comparing
      // we need to reset if any of the streams are different
      const getStreamId = (stream: AirbyteStreamAndConfiguration) => {
        return `${stream.stream?.namespace ?? ""}-${stream.stream?.name}`;
      };
      const lookupConnectionValuesStreamById = connection.syncCatalog.streams.reduce<
        Record<string, AirbyteStreamAndConfiguration>
      >((agg, stream) => {
        agg[getStreamId(stream)] = stream;
        return agg;
      }, {});
      const hasUserChangesInEnabledStreamsRequiringReset = values.syncCatalog.streams.some((_stream) => {
        const formStream = structuredClone(_stream);
        const connectionStream = structuredClone(lookupConnectionValuesStreamById[getStreamId(formStream)]);

        return !compareObjectsByFields<AirbyteStreamConfiguration>(formStream.config, connectionStream.config, [
          "cursorField",
          "destinationSyncMode",
          "primaryKey",
          "selectedFields",
          "syncMode",
          "aliasName",
        ]);
      });

      const catalogHasChanged = hasDiffInEnabledStream || hasUserChangesInEnabledStreams;
      const catalogChangesRequireReset = hasDiffInEnabledStream || hasUserChangesInEnabledStreamsRequiringReset;

      setSubmitError(null);

      // Whenever the catalog changed show a warning to the user, that we're about to reset their data.
      // Given them a choice to opt-out in which case we'll be sending skipReset: true to the update
      // endpoint.
      try {
        if (catalogChangesRequireReset) {
          const stateType = await getStateType(connection.connectionId);
          const result = await openModal<boolean>({
            title: formatMessage({ id: "connection.resetModalTitle" }),
            size: "md",
            content: (props) => <ResetWarningModal {...props} stateType={stateType} />,
          });
          if (result.type !== "canceled") {
            // Save the connection taking into account the correct skipReset value from the dialog choice.
            await saveConnection(values, {
              skipReset: !result.reason,
              catalogHasChanged,
            });
          } else {
            // We don't want to set saved to true or schema has been refreshed to false.
            return Promise.reject();
          }
        } else {
          // The catalog hasn't changed, or only added/removed stream(s). We don't need to ask for any confirmation and can simply save.
          await saveConnection(values, { skipReset: true, catalogHasChanged });
        }
        return Promise.resolve();
      } catch (e) {
        setSubmitError(e);
      }
    },
    [
      connection.catalogDiff?.transforms,
      connection.syncCatalog.streams,
      connection.connectionId,
      setSubmitError,
      getStateType,
      openModal,
      formatMessage,
      saveConnection,
    ]
  );

  useConfirmCatalogDiff();

  useUnmount(() => {
    discardRefreshedSchema();
  });

  const { state } = useLocation();
  useEffect(() => {
    if (typeof state === "object" && state && "triggerRefreshSchema" in state && state.triggerRefreshSchema) {
      refreshSchema();
    }
  }, [refreshSchema, state]);

  return (
    <FlexContainer direction="column" className={styles.content}>
      {schemaError && !schemaRefreshing ? (
        <SchemaError schemaError={schemaError} />
      ) : !schemaRefreshing && connection ? (
        <Form<FormConnectionFormValues>
          defaultValues={initialValues}
          schema={validationSchema}
          onSubmit={onFormSubmit}
          trackDirtyChanges
          disabled={mode === "readonly"}
        >
          <FlexContainer direction="column">
            <SchemaChangeMessage />
            <SchemaChangeBackdrop>
              <ConnectionConfigurationCard />
              <SyncCatalogCard />
              <div className={styles.editControlsContainer}>
                <UpdateConnectionFormControls onCancel={discardRefreshedSchema} />
              </div>
            </SchemaChangeBackdrop>
          </FlexContainer>
        </Form>
      ) : (
        <LoadingSchema />
      )}
    </FlexContainer>
  );
};
