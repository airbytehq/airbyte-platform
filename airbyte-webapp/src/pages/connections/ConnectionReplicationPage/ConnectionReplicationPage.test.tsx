import { render as tlr, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React, { Suspense } from "react";
import { VirtuosoMockContext } from "react-virtuoso";

import { mockConnection } from "test-utils/mock-data/mockConnection";
import {
  mockDestinationDefinition,
  mockDestinationDefinitionSpecification,
  mockDestinationDefinitionVersion,
} from "test-utils/mock-data/mockDestination";
import {
  mockSourceDefinition,
  mockSourceDefinitionSpecification,
  mockSourceDefinitionVersion,
} from "test-utils/mock-data/mockSource";
import { mockTheme } from "test-utils/mock-data/mockTheme";
import { mockWorkspace } from "test-utils/mock-data/mockWorkspace";
import { mockWorkspaceId } from "test-utils/mock-data/mockWorkspaceId";
import { TestWrapper, useMockIntersectionObserver } from "test-utils/testutils";

import { useGetConnectionQuery } from "core/api";
import { WebBackendConnectionUpdate } from "core/api/types/AirbyteClient";
import { defaultOssFeatures, FeatureItem } from "core/services/features";
import { ConnectionEditServiceProvider } from "hooks/services/ConnectionEdit/ConnectionEditService";

import { ConnectionReplicationPage } from "./ConnectionReplicationPage";

jest.setTimeout(40000);

jest.mock("area/workspace/utils", () => ({
  useCurrentWorkspaceId: () => mockWorkspaceId,
}));

jest.mock("core/api", () => ({
  useCurrentWorkspace: () => mockWorkspace,
  useGetConnectionQuery: jest.fn(() => async () => mockConnection),
  useGetConnection: () => mockConnection,
  useGetStateTypeQuery: () => async () => "global",
  useUpdateConnection: () => ({
    mutateAsync: async (connection: WebBackendConnectionUpdate) => connection,
    isLoading: false,
  }),
  useSourceDefinitionVersion: () => mockSourceDefinitionVersion,
  useDestinationDefinitionVersion: () => mockDestinationDefinitionVersion,
  useGetSourceDefinitionSpecification: () => mockSourceDefinitionSpecification,
  useGetDestinationDefinitionSpecification: () => mockDestinationDefinitionSpecification,
  useSourceDefinition: () => mockSourceDefinition,
  useDestinationDefinition: () => mockDestinationDefinition,
  LogsRequestError: jest.requireActual("core/api/errors").LogsRequestError,
}));

jest.mock("hooks/theme/useAirbyteTheme", () => ({
  useAirbyteTheme: () => mockTheme,
}));

describe("ConnectionReplicationPage", () => {
  const Wrapper: React.FC<React.PropsWithChildren<unknown>> = ({ children }) => (
    <Suspense fallback={<div>I should not show up in a snapshot</div>}>
      <TestWrapper>
        <ConnectionEditServiceProvider connectionId={mockConnection.connectionId}>
          <VirtuosoMockContext.Provider value={{ viewportHeight: 1000, itemHeight: 50 }}>
            {children}
          </VirtuosoMockContext.Provider>
        </ConnectionEditServiceProvider>
      </TestWrapper>
    </Suspense>
  );

  const render = async () => {
    let renderResult: ReturnType<typeof tlr>;

    await act(async () => {
      renderResult = tlr(
        <Wrapper>
          <ConnectionReplicationPage />
        </Wrapper>
      );
    });
    return renderResult!;
  };

  const setupSpies = (getConnection?: () => Promise<void>) => {
    (useGetConnectionQuery as jest.Mock).mockImplementation(() => getConnection);
  };

  beforeEach(() => {
    useMockIntersectionObserver();
  });

  it("should render", async () => {
    setupSpies();

    const renderResult = await render();
    expect(renderResult).toMatchSnapshot();
  });

  it("should show an error if there is a schemaError", async () => {
    setupSpies(() => Promise.reject("Test Error"));

    const renderResult = await render();

    await act(async () => {
      renderResult.queryByText("Refresh source schema")?.click();
    });
    expect(renderResult).toMatchSnapshot();
  });

  it("should show loading if the schema is refreshing", async () => {
    // Return pending promise
    setupSpies(() => new Promise(() => null));

    const renderResult = await render();
    await act(async () => {
      renderResult.queryByText("Refresh source schema")?.click();
    });

    await act(async () => {
      expect(renderResult.findByText("We are fetching the schema of your data source.", { exact: false })).toBeTruthy();
    });
  });

  describe("cron expression validation", () => {
    const INVALID_CRON_EXPRESSION = "invalid cron expression";
    const CRON_EXPRESSION_EVERY_MINUTE = "* * * * * * ?";

    it("should display an error for an invalid cron expression", async () => {
      setupSpies();
      const renderResult = await render();

      await userEvent.click(renderResult.getByTestId("configuration-card-expand-arrow"));

      await userEvent.click(renderResult.getByTestId("schedule-type-listbox-button"));
      await userEvent.click(renderResult.getByTestId("cron-option"));

      const cronExpressionInput = renderResult.getByTestId("cronExpression");

      await userEvent.clear(cronExpressionInput);
      await userEvent.type(cronExpressionInput, INVALID_CRON_EXPRESSION, { delay: 1 });

      const errorMessage = renderResult.getByText(/must contain at least 6 fields/);

      expect(errorMessage).toBeInTheDocument();
    });

    it("should allow cron expressions under one hour when feature enabled", async () => {
      setupSpies();

      const renderResult = await render();

      await userEvent.click(renderResult.getByTestId("configuration-card-expand-arrow"));

      await userEvent.click(renderResult.getByTestId("schedule-type-listbox-button"));
      await userEvent.click(renderResult.getByTestId("cron-option"));

      const cronExpressionField = renderResult.getByTestId("cronExpression");

      await userEvent.clear(cronExpressionField);
      await userEvent.type(cronExpressionField, CRON_EXPRESSION_EVERY_MINUTE, { delay: 1 });

      const errorMessage = renderResult.queryByTestId("cronExpressionError");

      expect(errorMessage).not.toBeInTheDocument();
    });

    it("should not allow cron expressions under one hour when feature not enabled", async () => {
      setupSpies();

      const featuresToInject = defaultOssFeatures.filter((f) => f !== FeatureItem.AllowSyncSubOneHourCronExpressions);

      const container = tlr(
        <TestWrapper features={featuresToInject}>
          <ConnectionEditServiceProvider connectionId={mockConnection.connectionId}>
            <ConnectionReplicationPage />
          </ConnectionEditServiceProvider>
        </TestWrapper>
      );

      await userEvent.click(container.getByTestId("configuration-card-expand-arrow"));

      await userEvent.click(container.getByTestId("schedule-type-listbox-button"));
      await userEvent.click(container.getByTestId("cron-option"));

      const cronExpressionField = container.getByTestId("cronExpression");

      await userEvent.clear(cronExpressionField);
      await userEvent.type(cronExpressionField, CRON_EXPRESSION_EVERY_MINUTE, { delay: 1 });

      const errorMessage = container.getByTestId("cronExpressionError");

      expect(errorMessage).toBeInTheDocument();
    });
  });
});
