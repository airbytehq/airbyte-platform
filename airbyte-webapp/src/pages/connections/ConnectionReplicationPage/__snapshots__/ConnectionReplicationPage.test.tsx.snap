// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionReplicationPage should render 1`] = `
<body>
  <div>
    <div
      class="<removed-for-snapshot-test>"
    >
      <form
        class="<removed-for-snapshot-test>"
      >
        <fieldset
          class="<removed-for-snapshot-test>"
        >
          <div
            class="<removed-for-snapshot-test>"
          >
            <div
              class="<removed-for-snapshot-test>"
            >
              <div
                class="<removed-for-snapshot-test>"
              >
                <div
                  class="<removed-for-snapshot-test>"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <div
                      class="<removed-for-snapshot-test>"
                    >
                      <div
                        class="<removed-for-snapshot-test>"
                      >
                        <h2
                          class="<removed-for-snapshot-test>"
                        >
                          Select streams
                        </h2>
                        <button
                          class="<removed-for-snapshot-test>"
                          data-testid="refresh-source-schema-btn"
                          type="button"
                        >
                          <div
                            class="<removed-for-snapshot-test>"
                            data-icon="sync"
                            data-testid="mocksvg"
                          />
                          <span
                            class="<removed-for-snapshot-test>"
                          >
                            Refresh source schema
                          </span>
                        </button>
                      </div>
                    </div>
                    <div
                      class="<removed-for-snapshot-test>"
                    >
                      <div
                        class="<removed-for-snapshot-test>"
                      >
                        <div
                          class="<removed-for-snapshot-test>"
                        >
                          <div
                            class="<removed-for-snapshot-test>"
                            data-testid="input-container"
                          >
                            <input
                              class="<removed-for-snapshot-test>"
                              data-testid="input"
                              placeholder="Search stream name"
                            />
                          </div>
                        </div>
                        <div
                          class="<removed-for-snapshot-test>"
                        >
                          <label
                            class="<removed-for-snapshot-test>"
                          >
                            <input
                              aria-checked="false"
                              class="<removed-for-snapshot-test>"
                              data-testid="hideDisableStreams-switch"
                              id="toggle-disabled-streams"
                              type="checkbox"
                              value=""
                            />
                            <span
                              class="<removed-for-snapshot-test>"
                            />
                          </label>
                          <label
                            for="toggle-disabled-streams"
                          >
                            <p
                              class="<removed-for-snapshot-test>"
                              data-type="text"
                            >
                              Hide disabled streams
                            </p>
                          </label>
                        </div>
                      </div>
                      <div
                        class="<removed-for-snapshot-test>"
                        data-testid="catalog-tree-table-body"
                      >
                        <div
                          class="<removed-for-snapshot-test>"
                          data-testid="catalog-tree-table-header"
                        >
                          <div
                            class="<removed-for-snapshot-test>"
                          >
                            <label
                              class="<removed-for-snapshot-test>"
                            >
                              <input
                                aria-checked="mixed"
                                class="<removed-for-snapshot-test>"
                                data-testid="all-streams-sync-switch"
                                id="all-streams-sync-switch"
                                type="checkbox"
                                value=""
                              />
                              <span
                                class="<removed-for-snapshot-test>"
                              />
                            </label>
                            <p
                              class="<removed-for-snapshot-test>"
                              data-type="text"
                            >
                              Sync
                            </p>
                          </div>
                          <div
                            class="<removed-for-snapshot-test>"
                          >
                            <p
                              class="<removed-for-snapshot-test>"
                              data-type="text"
                            >
                              Namespace
                              <button
                                class="<removed-for-snapshot-test>"
                                type="button"
                              >
                                <span
                                  class="<removed-for-snapshot-test>"
                                >
                                  <div
                                    class="<removed-for-snapshot-test>"
                                    data-icon="gear"
                                    data-testid="mocksvg"
                                  />
                                </span>
                              </button>
                            </p>
                          </div>
                          <div
                            class="<removed-for-snapshot-test>"
                          >
                            <p
                              class="<removed-for-snapshot-test>"
                              data-type="text"
                            >
                              Stream
                              <button
                                class="<removed-for-snapshot-test>"
                                type="button"
                              >
                                <span
                                  class="<removed-for-snapshot-test>"
                                >
                                  <div
                                    class="<removed-for-snapshot-test>"
                                    data-icon="gear"
                                    data-testid="mocksvg"
                                  />
                                </span>
                              </button>
                            </p>
                          </div>
                          <div
                            class="<removed-for-snapshot-test>"
                          >
                            <p
                              class="<removed-for-snapshot-test>"
                              data-type="text"
                            >
                              Sync mode
                              <span
                                class="<removed-for-snapshot-test>"
                              >
                                <span
                                  class="<removed-for-snapshot-test>"
                                >
                                  <span
                                    class="<removed-for-snapshot-test>"
                                  >
                                    <div
                                      class="<removed-for-snapshot-test>"
                                      data-icon="info-outline"
                                      data-testid="mocksvg"
                                    />
                                  </span>
                                </span>
                              </span>
                            </p>
                          </div>
                          <div
                            class="<removed-for-snapshot-test>"
                          >
                            <p
                              class="<removed-for-snapshot-test>"
                              data-type="text"
                            />
                          </div>
                          <div
                            class="<removed-for-snapshot-test>"
                          >
                            <p
                              class="<removed-for-snapshot-test>"
                              data-type="text"
                            >
                              Fields
                            </p>
                          </div>
                        </div>
                        <div
                          data-virtuoso-scroller="true"
                          style="position: relative; height: 150px;"
                        >
                          <div
                            data-viewport-type="window"
                            style="width: 100%; height: 100%; position: absolute; top: 0px;"
                          >
                            <div
                              data-testid="virtuoso-item-list"
                              style="box-sizing: border-box; padding-top: 0px; padding-bottom: 0px; margin-top: 0px;"
                            >
                              <div
                                data-index="0"
                                data-item-index="0"
                                data-known-size="50"
                              >
                                <div
                                  class="<removed-for-snapshot-test>"
                                  data-testid="catalog-tree-table-row-no-namespace-another_stream"
                                >
                                  <div
                                    class="<removed-for-snapshot-test>"
                                    data-noexpand="true"
                                  >
                                    <label
                                      class="<removed-for-snapshot-test>"
                                    >
                                      <input
                                        aria-checked="true"
                                        checked=""
                                        class="<removed-for-snapshot-test>"
                                        data-testid="selected-switch"
                                        type="checkbox"
                                        value=""
                                      />
                                      <span
                                        class="<removed-for-snapshot-test>"
                                      />
                                    </label>
                                    <div
                                      class="<removed-for-snapshot-test>"
                                    />
                                  </div>
                                  <div
                                    class="<removed-for-snapshot-test>"
                                    data-testid="destination-namespace-cell"
                                  >
                                    <span
                                      class="<removed-for-snapshot-test>"
                                    >
                                      <p
                                        class="<removed-for-snapshot-test>"
                                        data-type="text"
                                      >
                                        &lt;destination schema&gt;
                                      </p>
                                    </span>
                                  </div>
                                  <div
                                    class="<removed-for-snapshot-test>"
                                    data-testid="destination-stream-name-cell"
                                  >
                                    <span
                                      class="<removed-for-snapshot-test>"
                                    >
                                      <p
                                        class="<removed-for-snapshot-test>"
                                        data-type="text"
                                      >
                                        another_stream
                                      </p>
                                    </span>
                                  </div>
                                  <div
                                    class="<removed-for-snapshot-test>"
                                  >
                                    <div
                                      data-testid="sync-mode-select"
                                    >
                                      <button
                                        aria-expanded="false"
                                        aria-haspopup="listbox"
                                        class="<removed-for-snapshot-test>"
                                        data-headlessui-state=""
                                        data-testid="sync-mode-select-listbox-button"
                                        id="headlessui-listbox-button-:r2:"
                                        type="button"
                                      >
                                        <div
                                          class="<removed-for-snapshot-test>"
                                          data-error="false"
                                          data-testid="pill-select-button"
                                        >
                                          <div
                                            class="<removed-for-snapshot-test>"
                                          >
                                            <p
                                              class="<removed-for-snapshot-test>"
                                              data-type="text"
                                            >
                                              Full refresh | Overwrite
                                            </p>
                                          </div>
                                          <div
                                            class="<removed-for-snapshot-test>"
                                            data-icon="caret-down"
                                            data-testid="mocksvg"
                                          />
                                        </div>
                                      </button>
                                      <div
                                        style="position: absolute; z-index: 9999; top: 0px; left: 0px; transform: translate(0px,5px);"
                                      />
                                    </div>
                                  </div>
                                  <div
                                    class="<removed-for-snapshot-test>"
                                  />
                                  <div
                                    class="<removed-for-snapshot-test>"
                                  >
                                    <div
                                      class="<removed-for-snapshot-test>"
                                      data-testid="field-selection-status"
                                    >
                                      <div
                                        class="<removed-for-snapshot-test>"
                                      >
                                        <p
                                          class="<removed-for-snapshot-test>"
                                          data-testid="field-selection-status-selected-count"
                                          data-type="text"
                                        >
                                          All
                                        </p>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                              <div
                                data-index="1"
                                data-item-index="1"
                                data-known-size="50"
                              >
                                <div
                                  class="<removed-for-snapshot-test>"
                                  data-testid="catalog-tree-table-row-no-namespace-pokemon"
                                >
                                  <div
                                    class="<removed-for-snapshot-test>"
                                    data-noexpand="true"
                                  >
                                    <label
                                      class="<removed-for-snapshot-test>"
                                    >
                                      <input
                                        aria-checked="true"
                                        checked=""
                                        class="<removed-for-snapshot-test>"
                                        data-testid="selected-switch"
                                        type="checkbox"
                                        value=""
                                      />
                                      <span
                                        class="<removed-for-snapshot-test>"
                                      />
                                    </label>
                                    <div
                                      class="<removed-for-snapshot-test>"
                                    />
                                  </div>
                                  <div
                                    class="<removed-for-snapshot-test>"
                                    data-testid="destination-namespace-cell"
                                  >
                                    <span
                                      class="<removed-for-snapshot-test>"
                                    >
                                      <p
                                        class="<removed-for-snapshot-test>"
                                        data-type="text"
                                      >
                                        &lt;destination schema&gt;
                                      </p>
                                    </span>
                                  </div>
                                  <div
                                    class="<removed-for-snapshot-test>"
                                    data-testid="destination-stream-name-cell"
                                  >
                                    <span
                                      class="<removed-for-snapshot-test>"
                                    >
                                      <p
                                        class="<removed-for-snapshot-test>"
                                        data-type="text"
                                      >
                                        pokemon
                                      </p>
                                    </span>
                                  </div>
                                  <div
                                    class="<removed-for-snapshot-test>"
                                  >
                                    <div
                                      data-testid="sync-mode-select"
                                    >
                                      <button
                                        aria-expanded="false"
                                        aria-haspopup="listbox"
                                        class="<removed-for-snapshot-test>"
                                        data-headlessui-state=""
                                        data-testid="sync-mode-select-listbox-button"
                                        id="headlessui-listbox-button-:r5:"
                                        type="button"
                                      >
                                        <div
                                          class="<removed-for-snapshot-test>"
                                          data-error="false"
                                          data-testid="pill-select-button"
                                        >
                                          <div
                                            class="<removed-for-snapshot-test>"
                                          >
                                            <p
                                              class="<removed-for-snapshot-test>"
                                              data-type="text"
                                            >
                                              Full refresh | Overwrite
                                            </p>
                                          </div>
                                          <div
                                            class="<removed-for-snapshot-test>"
                                            data-icon="caret-down"
                                            data-testid="mocksvg"
                                          />
                                        </div>
                                      </button>
                                      <div
                                        style="position: absolute; z-index: 9999; top: 0px; left: 0px; transform: translate(0px,5px);"
                                      />
                                    </div>
                                  </div>
                                  <div
                                    class="<removed-for-snapshot-test>"
                                  />
                                  <div
                                    class="<removed-for-snapshot-test>"
                                  >
                                    <div
                                      class="<removed-for-snapshot-test>"
                                      data-testid="field-selection-status"
                                    >
                                      <div
                                        class="<removed-for-snapshot-test>"
                                      >
                                        <p
                                          class="<removed-for-snapshot-test>"
                                          data-testid="field-selection-status-selected-count"
                                          data-type="text"
                                        >
                                          All
                                        </p>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                              <div
                                data-index="2"
                                data-item-index="2"
                                data-known-size="50"
                              >
                                <div
                                  class="<removed-for-snapshot-test>"
                                  data-testid="catalog-tree-table-row-no-namespace-pokemon2"
                                >
                                  <div
                                    class="<removed-for-snapshot-test>"
                                    data-noexpand="true"
                                  >
                                    <label
                                      class="<removed-for-snapshot-test>"
                                    >
                                      <input
                                        aria-checked="false"
                                        class="<removed-for-snapshot-test>"
                                        data-testid="selected-switch"
                                        type="checkbox"
                                        value=""
                                      />
                                      <span
                                        class="<removed-for-snapshot-test>"
                                      />
                                    </label>
                                    <div
                                      class="<removed-for-snapshot-test>"
                                    />
                                  </div>
                                  <div
                                    class="<removed-for-snapshot-test>"
                                    data-testid="destination-namespace-cell"
                                  >
                                    <span
                                      class="<removed-for-snapshot-test>"
                                    >
                                      <p
                                        class="<removed-for-snapshot-test>"
                                        data-type="text"
                                      >
                                        &lt;destination schema&gt;
                                      </p>
                                    </span>
                                  </div>
                                  <div
                                    class="<removed-for-snapshot-test>"
                                    data-testid="destination-stream-name-cell"
                                  >
                                    <span
                                      class="<removed-for-snapshot-test>"
                                    >
                                      <p
                                        class="<removed-for-snapshot-test>"
                                        data-type="text"
                                      >
                                        pokemon2
                                      </p>
                                    </span>
                                  </div>
                                  <div
                                    class="<removed-for-snapshot-test>"
                                  />
                                  <div
                                    class="<removed-for-snapshot-test>"
                                  />
                                  <div
                                    class="<removed-for-snapshot-test>"
                                  />
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="<removed-for-snapshot-test>"
            >
              <div
                class="<removed-for-snapshot-test>"
              >
                <div
                  class="<removed-for-snapshot-test>"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <button
                      class="<removed-for-snapshot-test>"
                      data-testid="cancel-edit-button"
                      disabled=""
                      type="button"
                    >
                      <span
                        class="<removed-for-snapshot-test>"
                      >
                        Cancel
                      </span>
                    </button>
                    <button
                      class="<removed-for-snapshot-test>"
                      data-testid="save-edit-button"
                      disabled=""
                      type="submit"
                    >
                      <span
                        class="<removed-for-snapshot-test>"
                      >
                        Save changes
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </fieldset>
      </form>
    </div>
  </div>
  <div
    class="<removed-for-snapshot-test>"
  />
</body>
`;

exports[`ConnectionReplicationPage should show an error if there is a schemaError 1`] = `
<body>
  <div>
    <div
      class="<removed-for-snapshot-test>"
    >
      <div
        class="<removed-for-snapshot-test>"
      >
        <div
          class="<removed-for-snapshot-test>"
        >
          <div
            class="<removed-for-snapshot-test>"
          >
            <div
              class="<removed-for-snapshot-test>"
            >
              <div
                class="<removed-for-snapshot-test>"
              >
                <button
                  class="<removed-for-snapshot-test>"
                  type="button"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                    data-icon="reset"
                    data-testid="mocksvg"
                  />
                  <span
                    class="<removed-for-snapshot-test>"
                  >
                    Try again
                  </span>
                </button>
              </div>
              <div
                class="<removed-for-snapshot-test>"
              >
                <div
                  class="<removed-for-snapshot-test>"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <div
                      class="<removed-for-snapshot-test>"
                      data-icon="status-error"
                      data-testid="mocksvg"
                    />
                  </div>
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <span
                      class="<removed-for-snapshot-test>"
                    >
                      Sorry. Something went wrong…
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="<removed-for-snapshot-test>"
  />
</body>
`;
